### PR TITLE
Redact access token from output

### PR DESCRIPTION
**Pull request by hemarina**
Fix internal issue #1449.
Created `RedactSensitiveData` to redact access token from output when running `--debug`

### CODE REVIEW

File: cli/azd/pkg/executil/executil.go [hunk 1]
```diff
@@ -197,10 +198,31 @@ type RunArgs struct {
 	EnrichError bool
 }
 
+type RedactData struct {
```
Comment (vhvb1989) [inline cli/azd/pkg/executil/executil.go:201]: Do we need to export this type?

```suggestion
type redactData struct {
```

File: cli/azd/pkg/executil/executil.go [hunk 2]
```diff
@@ -197,10 +198,31 @@ type RunArgs struct {
...
+	matchString   *regexp.Regexp
+	replaceString string
+}
+
 func (rr RunResult) String() string {
 	return fmt.Sprintf("exit code: %d, stdout: %s, stderr: %s", rr.ExitCode, rr.Stdout, rr.Stderr)
 }
 
+func RedactSensitiveData(msg string) string {
```
Comment (vhvb1989) [inline cli/azd/pkg/executil/executil.go:210]: Same here. 
Right now, this function is only used by `RunWithResult()`, so we don't need to export it.

We can consider creating a Redactor package in the future, which would take any `RedactData` and do the sanitization.  But that can be done in the future.

File: cli/azd/pkg/executil/executil.go [hunk 3]
```diff
@@ -197,10 +198,31 @@ type RunArgs struct {
...
+	regexpRedact := make(map[string]RedactData)
```
Comment (vhvb1989) [inline cli/azd/pkg/executil/executil.go:211]: consider a name change here like:

```suggestion
	regexpRedactRules := make(map[string]RedactData)
```

File: cli/azd/pkg/executil/executil.go [hunk 4]
```diff
@@ -197,10 +198,31 @@ type RunArgs struct {
...
+	regexpRedact["access token"] = RedactData{
+		matchString:   regexp.MustCompile("\"accessToken\": \".*\""),
+		replaceString: "\"accessToken\": \"redact to prevent sensitive data\"",
+	}
+
+	for _, val := range regexpRedact {
```
Comment (vhvb1989) [inline cli/azd/pkg/executil/executil.go:217]:
```suggestion
	for _, redactRule := range regexpRedact {
```

File: cli/azd/pkg/executil/executil.go [hunk 5]
```diff
@@ -197,10 +198,31 @@ type RunArgs struct {
...
+		regMatchString := val.matchString.FindString(msg)
+		if regMatchString != "" && strings.Contains(msg, regMatchString) {
+			return strings.Replace(msg, regMatchString, val.replaceString, -1)
```
Comment (vhvb1989) [inline cli/azd/pkg/executil/executil.go:220]: can we use [ReplaceAllString](https://pkg.go.dev/regexp#Regexp.ReplaceAllString) ??  I think we can avoid creating the `regMatchString`

**Review by vhvb1989 — COMMENTED** [review]
Can you add unit tests for RedactSensitiveData ?

At least 3 cases:
- When msg contains an access token and validate it is updated
- When msg does not contain access token and validate the msg is not changed
- When the msg contains more than one access token and validate all instances are updated

Thank you.
This is a good foundation that allows adding more `RedactData` items

File: cli/azd/pkg/executil/executil.go [hunk 6]
```diff
@@ -258,7 +280,7 @@ func RunWithResult(ctx context.Context, args RunArgs) (RunResult, error) {
 	err = cmd.Wait()
 
 	if args.Debug {
-		log.Printf("Exit Code:%d\nOut:%s\nErr:%s\n", cmd.ProcessState.ExitCode(), stdout.String(), stderr.String())
+		log.Printf("Exit Code:%d\nOut:%s\nErr:%s\n", cmd.ProcessState.ExitCode(), redactSensitiveData(stdout.String()), stderr.String())
```
Comment (ellismg) [inline cli/azd/pkg/executil/executil.go:283]: Could we also run `redactSensitiveData` over `stderr.String()` as well?

File: cli/azd/pkg/executil/executil.go [hunk 7]
```diff
@@ -197,10 +198,31 @@ type RunArgs struct {
...
+	matchString   *regexp.Regexp
+	replaceString string
+}
+
 func (rr RunResult) String() string {
 	return fmt.Sprintf("exit code: %d, stdout: %s, stderr: %s", rr.ExitCode, rr.Stdout, rr.Stderr)
 }
 
+func redactSensitiveData(msg string) string {
+	regexpRedactRules := make(map[string]redactData)
```
Comment (ellismg) [inline cli/azd/pkg/executil/executil.go:211]: You should be able to use a map literal here (instead of doing a make + an assign).  See https://go.dev/tour/moretypes/20 for an example.

**Review by ellismg — APPROVED** [review]

File: cli/azd/pkg/executil/executil.go [hunk 8]
```diff
@@ -197,10 +198,29 @@ type RunArgs struct {
...
+
 func (rr RunResult) String() string {
 	return fmt.Sprintf("exit code: %d, stdout: %s, stderr: %s", rr.ExitCode, rr.Stdout, rr.Stderr)
 }
 
+func redactSensitiveData(msg string) string {
+	var regexpRedactRules = map[string]redactData{
+		"access token": {
+			regexp.MustCompile("\"accessToken\": \".*\""),
+			"\"accessToken\": \"redact to prevent sensitive data\"",
```
Comment (weikanglim) [inline cli/azd/pkg/executil/executil.go:214]: For correct grammar, I think this sentence would need to say "redact**ed** to prevent sensitive data **leakage**". To avoid the long sentence, I think just printing `<redacted>` would suffice 

```suggestion
			"\"accessToken\": \"<redacted>\"",
```

File: cli/azd/pkg/executil/executil_test.go [hunk 9]
```diff
@@ -161,3 +161,54 @@ func TestRunEnrichError(t *testing.T) {
 	// themselves in checking the ExitCode.
 	require.EqualError(t, err, fmt.Sprintf("%s: exit status 2", res.String()))
 }
+
+func TestRedactSensitiveData(t *testing.T) {
```
Comment (weikanglim) [inline cli/azd/pkg/executil/executil_test.go:165]: Just pointing out you could write the three test cases here as a "table-driven style" test, where you set up all your expected input/output before-hand, and write a for-loop that will execute each test matching input to output:

Here's a documented [example](https://gobyexample.com/testing).

Pseudocode related to your test case:

```golang
tests := []struct{
  scenario string,
  input string,
  expected string) {
  {scenario: "Basic",
   input: `"accessToken": "eyJ0eX",
"expiresOn": "2022-08-11 10:33:39.000000",
"subscription": "2cd61",
"tenant": "72f988bf",
"tokenType": "Bearer"
}`,
  expected: `"accessToken": "redact to prevent sensitive data",
"expiresOn": "2022-08-11 10:33:39.000000",
"subscription": "2cd61",
"tenant": "72f988bf",
"tokenType": "Bearer"
}`},
  {scenario: "NoReplacement,", ... },
  {scenario: "MultipleReplacement,", ... },
}

for _, test := range tests {
  t.Run(test.scenario, func() {
    actual := redactSensitiveData(test.input)
    require.Equal(t, test.expected, actual)
  })
}
```

**Review by weikanglim — APPROVED** [review]
LGTM, minor comments

File: cli/azd/cmd/infra_create.go [hunk 10]
```diff
@@ -81,7 +80,7 @@ func (ica *infraCreateAction) Run(ctx context.Context, cmd *cobra.Command, args
 
 	// Copy the parameter template file to the environment working directory and do substitutions.
 	parametersPath := azdCtx.BicepParametersTemplateFilePath(rootModule)
-	parametersBytes, err := ioutil.ReadFile(parametersPath)
+	parametersBytes, err := os.ReadFile(parametersPath)
```
Comment (vhvb1989) [inline cli/azd/cmd/infra_create.go:83]: curious on why this change, from ioutil to os, is part of the PR??

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/cmd/infra_create.go (hunk 10 above)
Comment (vhvb1989) [reply cli/azd/cmd/infra_create.go:83]: I would suggest creating a new issue about the lint error and a PR to fix that one, instead of using the same PR.
A fix for a specific issue deserves its own PR and git history entry, specially when we do squash-commit for merging PRs,
which means that there will be just one commit for your new feature to `redact access token`

No need to change or do it on this one, but keep in mind in case there's a case like this in the future.
